[Site Isolation] change downcast to dynamicDowncast in JSDOMBindingSecurity::remoteFrameAccessError

atar13 opened the pull request:
#### 8fb7be5683f43a48a723b6a950bfe6268fe02118
<pre>
[Site Isolation] change downcast to dynamicDowncast in JSDOMBindingSecurity::remoteFrameAccessError
<a href="https://bugs.webkit.org/show_bug.cgi?id=297874">https://bugs.webkit.org/show_bug.cgi?id=297874</a>
<a href="https://rdar.apple.com/159134034">rdar://159134034</a>

Reviewed by Alex Christensen.

Change a downcast to a dynamicDowncast to handle cases where an invalid downcast between a RemoteDOMWindow to a LocalDOMWindow was attempted. Previously, if the cast failed, an assert was tripped and caused a crash.

See Bugzilla for steps to reproduce. The crash only occurred with Site Isolation enabled.

* Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
(WebCore::remoteFrameAccessError):

Canonical link: <a href="https://commits.webkit.org/299777@main">https://commits.webkit.org/299777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c70e65ee6fd36e3f95e441c44753d8285d0a622

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120137 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126486 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72210 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48408 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91239 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60544 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123089 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32373 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107714 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71790 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25820 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70112 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26007 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99857 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47423 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99700 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25313 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45154 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23176 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43688 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46919 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52625 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46387 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49734 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48071 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8dc60d40f94544612651e4f67ffac6f36c452dd6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49861

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -46,9 +50,9 @@ void printErrorMessageForFrame(LocalFrame* frame, const String& message)
 // FIXME: Refactor to share code with LocalDOMWindow::crossDomainAccessErrorMessage.
 static String remoteFrameAccessError(JSC::JSGlobalObject* lexicalGlobalObject)
```
We should probably pass in the RemoteDOMWindow from BindingSecurity::shouldAllowAccessToDOMWindow to reportErrorAccessingRemoteFrame, and use that to get the RemoteFrame, and use that to access frameDocumentSecurityOrigin, all with proper null checks.

atar13 replied:
Perfect. I wasn't sure where the securityOrigin was within RemoteDOMWindow, but I'll just get it from RemoteFrame.

atar13 replied:
Updated to use `frameDocumentSecurityOriginOrOpaque` in https://github.com/WebKit/WebKit/commit/597b71e21ae7f855ac476a5d768118787d19e9a2

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa1044faad6807df3b546c8fc5852e3ea0849239. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49861

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -44,21 +50,27 @@ void printErrorMessageForFrame(LocalFrame* frame, const String& message)
 
 // FIXME: Refactor to share code with LocalDOMWindow::crossDomainAccessErrorMessage.
-static String remoteFrameAccessError(JSC::JSGlobalObject* lexicalGlobalObject)
+static String remoteFrameAccessError(JSC::JSGlobalObject* lexicalGlobalObject, std::optional<RemoteFrame*> remoteFrame)
```
A RemoteFrame* is already optional.  We can check it for nullness.

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -44,21 +50,27 @@ void printErrorMessageForFrame(LocalFrame* frame, const String& message)
-    Ref activeOrigin = active.document()->securityOrigin();
-    return makeString("Blocked a frame with origin \""_s, activeOrigin->toString(), "\" from accessing a cross-origin frame. Protocols, domains, and ports must match."_s);
+    String activeOrigin;
+    if (remoteFrame.has_value())
```
I think this would look better as a ternary operator.

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -22,14 +22,20 @@
+#include "JSDOMWindowCustom.h"
 #include "LocalDOMWindow.h"
 #include "LocalFrame.h"
+#include "RemoteDOMWindow.h"
```
This is the only new include I think should be necessary.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/908421da06b386299460920733664fa67f391391. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49861

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/436c77f39a6bd6acb3d58d8b177b094c5cafb091. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49861

atar13 commented:
> I think this indicates an issue with the change. The frame is showing http://127.0.0.1:8000/site-isolation/resources/window-properties-grandchild.html when this error is generated, so the origin should be http://127.0.0.1:8000/

Good catch! I think I was grabbing the wrong RemoteFrame.

Before I was down casting `DOMWindow` to `RemoteDOMWindow` and getting the `RemoteFrame` from there.  Now I'm taking the `JSC::JSCGlobalObject` and casting to a `RemoteDOMWindow` using `asJSDOMWindow` and that seems to be giving the right origin of 127.0.0.1.

I've reverted the changes to the `-expected.txt` as well.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2bd0ecb97a42bb7086b538d428e78a0a9e8b1c2a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49861

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -46,9 +48,9 @@ void printErrorMessageForFrame(LocalFrame* frame, const String& message)
-    auto& active = activeDOMWindow(*lexicalGlobalObject);
-    Ref activeOrigin = active.document()->securityOrigin();
-    return makeString("Blocked a frame with origin \""_s, activeOrigin->toString(), "\" from accessing a cross-origin frame. Protocols, domains, and ports must match."_s);
+    auto* remoteWindow = dynamicDowncast<RemoteDOMWindow>(asJSDOMWindow(lexicalGlobalObject)->wrapped());
```
RefPtr remoteWindow = ...

achristensen07 commented on Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
```diff
@@ -46,9 +48,9 @@ void printErrorMessageForFrame(LocalFrame* frame, const String& message)
-    Ref activeOrigin = active.document()->securityOrigin();
-    return makeString("Blocked a frame with origin \""_s, activeOrigin->toString(), "\" from accessing a cross-origin frame. Protocols, domains, and ports must match."_s);
+    auto* remoteWindow = dynamicDowncast<RemoteDOMWindow>(asJSDOMWindow(lexicalGlobalObject)->wrapped());
+    const SecurityOrigin& activeOrigin = remoteWindow ? remoteWindow->frame()->frameDocumentSecurityOriginOrOpaque() : activeDOMWindow(*lexicalGlobalObject).document()->securityOrigin();
```
Ref activeOrigin = ...

atar13 committed:
[Site Isolation] change downcast to dynamicDowncast in JSDOMBindingSecurity::remoteFrameAccessError
https://bugs.webkit.org/show_bug.cgi?id=297874
rdar://159134034

Reviewed by Alex Christensen.

Change a downcast to a dynamicDowncast to handle cases where an invalid downcast between a RemoteDOMWindow to a LocalDOMWindow was attempted. Previously, if the cast failed, an assert was tripped and caused a crash.

See Bugzilla for steps to reproduce. The crash only occurred with Site Isolation enabled.

* Source/WebCore/bindings/js/JSDOMBindingSecurity.cpp:
(WebCore::remoteFrameAccessError):

Canonical link: https://commits.webkit.org/299777@main
